pkg/volumemanager/types.go

```diff
@@ -0,0 +1,173 @@
+	for i, obj := range objs {
+		volumes[i] = obj.(*api.PersistentVolume)
+	}
+
+	comparator := persistentVolumeComparator{volumes}
+	sort.Sort(comparator)
+	return comparator.volumes, nil
+}
+
+// MatchPredicate is a function that indicates that a persistent volume matches another
+type MatchPredicate func(compareThis, toThis *api.PersistentVolume) bool
+
+// Find returns the nearest PV from the ordered list or nil if a match is not found
+func (pvstore *persistentVolumeOrderedIndex) Find(pv *api.PersistentVolume, matchPredicate MatchPredicate) (*api.PersistentVolume, error) {
+	volumes, err := pvstore.ListByAccessModes(pv.Spec.AccessModes)
+	if err != nil {
+		return nil, err
+	}
+	i := sort.Search(len(volumes), func(i int) bool { return matchPredicate(pv, volumes[i]) })
+	if i < len(volumes) {
+		return volumes[i], nil
+	}
+	return nil, nil
+}
+
+// FindByStorageCapacity is a convenience method that calls Find w/ requisite MatchPredicate for storage
+func (pvstore *persistentVolumeOrderedIndex) FindByAccessModesAndStorageCapacity(modes []api.AccessModeType, qty resource.Quantity) (*api.PersistentVolume, error) {
+	pv := &api.PersistentVolume{
+		Spec: api.PersistentVolumeSpec{
+			AccessModes: modes,
+			Capacity: api.ResourceList{
+				api.ResourceName(api.ResourceStorage): qty,
+			},
+		},
+	}
+	return pvstore.Find(pv, onlyUnboundedVolumes)
+}
+
+// FindByClaim is a convenience method that finds a volume by the claim's AccessModes and requests for Storage
+func (pvstore *persistentVolumeOrderedIndex) FindByClaim(claim *api.PersistentVolumeClaim) (*api.PersistentVolume, error) {
+	return pvstore.FindByAccessModesAndStorageCapacity(claim.Spec.AccessModes, claim.Spec.Resources.Requests[api.ResourceName(api.ResourceStorage)])
+}
+
+// persistentVolumeComparator is used to order volumes by ascending storage size
+type persistentVolumeComparator struct {
+	volumes []*api.PersistentVolume
+}
+
+func (c persistentVolumeComparator) Less(i, j int) bool {
+	return matchStorageCapacity(c.volumes[i], c.volumes[j])
+}
+
+func (c persistentVolumeComparator) Swap(i, j int) {
+	c.volumes[i], c.volumes[j] = c.volumes[j], c.volumes[i]
+}
+
+func (c persistentVolumeComparator) Len() int {
+	return len(c.volumes)
+}
+
+// matchStorageCapacity is a MatchPredicate used to sort and find volumes
+func matchStorageCapacity(pvA, pvB *api.PersistentVolume) bool {
+	// skip already claimed volumes
+	if pvA.Spec.ClaimRef != nil {
+		return false
+	}
+
+	aQty := pvA.Spec.Capacity[api.ResourceStorage]
+	bQty := pvB.Spec.Capacity[api.ResourceStorage]
+	aSize := aQty.Value()
+	bSize := bQty.Value()
+	return aSize <= bSize
+}
+
+// onlyUnboundedVolumes is a MatchPredicate that filters bound volumes before comparing storage capacity
+func onlyUnboundedVolumes(compareThis, toThis *api.PersistentVolume) bool {
+	if toThis.Spec.ClaimRef != nil || compareThis.Spec.ClaimRef != nil {
+		return false
+	}
+	return matchStorageCapacity(compareThis, toThis)
```

Comment:
pattern-wise, this function is not self-describing, either rename to something like matchUnboundCapacity() or something.  Since it is only used in one place, you might just inline it.